packages/selected-network-controller/src/SelectedNetworkController.ts

```diff
@@ -192,21 +192,60 @@ export class SelectedNetworkController extends BaseController<
 
     this.messagingSystem.subscribe(
       'NetworkController:stateChange',
-      (availableNetworkClientIds) => {
-        // if a network is updated or removed, update the networkClientId for all domains
-        // that were using it to the selected network client id
-        const { selectedNetworkClientId } = this.messagingSystem.call(
-          'NetworkController:getState',
-        );
-        Object.entries(this.state.domains).forEach(
-          ([domain, networkClientIdForDomain]) => {
-            if (!availableNetworkClientIds.includes(networkClientIdForDomain)) {
-              this.setNetworkClientIdForDomain(domain, selectedNetworkClientId);
-            }
-          },
+      (
+        { selectedNetworkClientId, networkConfigurationsByChainId },
+        patches,
+      ) => {
+        const patch = patches.find(
+          ({ op, path }) =>
+            (op === 'replace' || op === 'remove') &&
+            path[0] === 'networkConfigurationsByChainId',
         );
+
+        if (patch) {
+          const networkClientIdToChainId = Object.values(
+            networkConfigurationsByChainId,
+          ).reduce((acc, network) => {
+            network.rpcEndpoints.forEach(
+              ({ networkClientId }) => (acc[networkClientId] = network.chainId),
+            );
+            return acc;
+          }, {} as Record<string, Hex>);
+
+          Object.entries(this.state.domains).forEach(
+            ([domain, networkClientIdForDomain]) => {
+              const chainIdForDomain =
+                networkClientIdToChainId[networkClientIdForDomain];
+
+              if (patch.op === 'remove' && !chainIdForDomain) {
+                // If the network was removed, fall back to the globally selected network
+                this.setNetworkClientIdForDomain(
+                  domain,
+                  selectedNetworkClientId,
+                );
```

Comment:
Wouldn't this set the chainId for the domain to the globally selected network if we just removed the rpcUrl that it was connected to? I would assume in this case we'd want to replace it with the new "default" rpcUrl for that chainId? And, if that network was removed entirely (not just the rpcUrl) we'd set it to the globally selected network?

Reply:
oh does the `remove` op only apply if the entire chainId is removed?

Reply:
Yes, the case you described falls into the replace operation